### DISCUSSION TITLE
Using Navigator takes a toll on the performance

### DISCUSSION THREAD

**Discussion by mwanagosos** [clientIO/joint#1692] [Q&A]
Hi. I've recently started using the Navigator feature
https://resources.jointjs.com/docs/rappid/v3.5/ui.Navigator.html

While it does the job, I've noticed that it makes scrolling a bit choppy when having a graph with a bigger number of shapes (about a 100). If I freeze the Navigator and hide it, the performance is a lot better.

I understand that using Navigator causes JointJS to have to update two separate papers and this is why the performance suffers a little. I was wondering if there are some techniques that could make this situation a bit better.

I'm using JointJS+ 3.5

```
new ui.PaperScroller({
  paper: this.paper,
  cursor: 'grab',
  autoResizePaper: true,
  scrollWhileDragging: true,
  baseHeight: 600,
  baseWidth: 600,
  contentOptions: {
    allowNewOrigin: 'any',
    useModelGeometry: true,
    padding: 50,
    minWidth: 1250,
    minHeight: 1250,
  },
});
```

```
new dia.Paper({
  model: this.graph,
  width: 1500,
  height: 1500,
  gridSize: 30,
  background: {
    color: '#ffffff',
  },
  drawGrid: {
    name: 'dot',
    thickness: 2,
    color: '#969696',
  },
  cellViewNamespace: cellNamespace,
  snapLinks: { radius: 75 },
  validateConnection(
    sourceCellView,
    sourceMagnet,
    targetCellView,
    targetMagnet,
  ) {
    // ...
  },
  validateMagnet(cellView, magnet) {
    // ...
  },
  defaultLink: new dia.Link({
    // ...
  }),
});
```

```
new ui.Navigator({
  paperScroller,
  width: 300,
  height: 200,
  padding: 10,
  zoomOptions: { max: 2, min: 0.2 },
  paperOptions: {
    cellViewNamespace: paper.options.cellViewNamespace,
  },
});
```

**Comment by kumilingus** [comment]
The `Navigator` affects performance because you are rendering the same graph twice. To improve, you can make the following adjustments:

1. use `dia.Paper.sorting.APPROX` for both paper and the navigator `sorting` paper [option](https://resources.jointjs.com/docs/jointjs/v3.5/joint.html#dia.Paper.prototype.options.sorting). (do not use legacy `EXACT` if not necessary, the `APPROX` sorting will became the default at some point)
2. make the navigator `async` (you can also specify a `batchSize` to lesser than the default `1000`).
3. use simplified custom element and link views as shown in [Kitchen Sink](https://resources.jointjs.com/demos/kitchensink) demo.

```js
new ui.Navigator({
    paperOptions: {
        // 1.
        sorting: dia.Paper.sorting.APPROX, // or dia.Paper.sorting.NONE
        // 2.
        async: true, // or { batchSize: 100 }, 
        // 3.
        elementView: MyElementView,
        linkView: MyLinkView,
        cellViewNamespace: { /* no other views are accessible in the navigator */ }
    }
});
```

**Reply by mwanagosos** [reply]
Thanks for the tips, I'll try them out.
The `batchSize` property does not seem to be a part of the `dia.Paper.Options` TypeScript interface though.

**Reply by kumilingus** [reply]
You are right, if you want to set the `batchSize` you have to unfreeze the navigator's paper.
```JS
navigator.unfreeze({ batchSize: 1 });
```

**Comment by mwanagosos** [comment]
Hi @kumilingus.
I would like to freeze the navigator and manually refresh its state - both the rendered elements and the zoom indicator. I think this might be a good way to improve the performance. Is there a way to do that?

**Reply by mwanagosos** [reply]
Currently, the navigator causes dragging elements and scrolling through the view to stutter quite a bit. Maybe debouncing it would improve this.

**Reply by kumilingus** [reply]
Here's how you disable specific navigator's features. I suggest you'd play with those, see what works best for you and get back to us and discuss what we can improve on our side.

### Current Viewport Frame
```js
// will stop listening to the `scroll` events.
navigator.options.paperScroller.$el.off(navigator.getEventNamespace());
// update the viewport frame position and size
navigator.updateCurrentView();
```

Note that `updateCurrentView` is already debounced (for 0ms). We can change it and make it optionally easily.

### Size of the Paper

```js
// stop listening to the `resize` event of the paper inside the paper scroller
navigator.stopListening();
// update the size of the paper
navigator.updatePaper()
```

### Paper Content
```js
// stop updating cells
navigator.freeze()
// update the cells and keep it frozen
navigator.targetPaper.dumpViews()
```

**Reply by mwanagosos** [reply]
The following properties are not listed in TypeScript:
- `navigator.options.paperScroller`
- `navigator.targetPaper`
Is there any way around that?

**Reply by kumilingus** [reply]
At this phase, we are only trying to find a solution, that we could implement directly in JointJS+.

You can ignore typescript errors in multiple ways. e.g.
```ts
navigator['targetPaper'].updatePaper();
```
```ts
// @ts-ignore
navigator.updatePaper();
```
Or edit `rappid.d.ts` and add the missing definitions. These methods were not intended to be public and are therefore not defined.

**Reply by mwanagosos** [reply]
I was wondering if there is a way to go around that in a nicer way. If not, I'll manage anyways :) thank you.

**Reply by mwanagosos** [reply]
@kumilingus I've noticed that there is a very big performance hit even when I do
```
paperScroller.$el.off(navigator.getEventNamespace());
navigator.stopListening();
navigator.freeze()
```
Scrolling through the graph drops from ~45fps to ~21fps if the navigator is in the DOM tree. Removing it by modifying the DOM tree manually causes the FPS to go back up.
I think having the navigator makes the DOM tree twice as big basically, and this might cause the performance issues.
Do you think I could improve that somehow?

**Reply by kumilingus** [reply]
You can use simplified views in the navigator. Similar what we done in the Kitchen Sink application.

**Reply by kumilingus** [reply]
Alternatively, the navigator could be implemented using the [SVGUseElement](https://developer.mozilla.org/en-US/docs/Web/SVG/Element/use) element (referencing the container of all `cells`). But the last time I tried that, it was incredibly slow. Perhaps it's time to re-test.
